File: RestroomFinder/Native/Android/RestroomFinder/app/src/main/java/org/actransit/restroomfinder/Infrastructure/ServerAPI.java
package org.actransit.restroomfinder.Infrastructure;


import org.actransit.restroomfinder.Model.FeedbackModel;
import org.actransit.restroomfinder.Model.OperationModel;
import org.actransit.restroomfinder.Model.RestroomModel;
import org.actransit.restroomfinder.Model.RouteModel;
import org.actransit.restroomfinder.Model.VersionModel;
import org.json.JSONObject;

import java.util.Calendar;
import java.util.List;
import android.content.Context;

/**
 * Created by atajadod on 5/16/2016.
 */
public class ServerAPI extends ServerBaseAPI{

    public ServerAPI(Context context){
        super(context);
    }

    public void getRestrooms(String route, Double latitude, Double longitude, ServerResult callBack){
        String url = getRestroomsURL(route==null?"":route,latitude,longitude,80000);
        Get(url, RestroomModel.class, callBack);
    }
    public void getRestroom(int restroomId, ServerResult callBack){
        String url = getRestroomURL(restroomId);
        Get(url, RestroomModel.class, callBack);
    }


    public void getRoutes(ServerResult callBack){
        String url = getRoutesURL();
        Get(url, RouteModel.class, callBack);
    }

    public void getOperation(String badge,boolean agreed,boolean  validating, ServerResult callBack){
        String surl = Constants.Server.operationUrl;
        OperationModel obj = new OperationModel();
        obj.badge = badge;
        obj.agreed = agreed;
        obj.validating = validating;
        obj.incidentDateTime = Calendar.getInstance().getTime().toString();
        JSONObject jObj=JSONMapper.toJSON(obj, OperationModel.class);
        Post(surl, jObj, OperationModel.class, callBack);
    }

    public void getFeedbacks(int restStopId, ServerResult callBack) {
        String url = getFeedbacksURL(restStopId);
        Get(url, FeedbackModel.class, callBack);
    }
    public void getVersion(ServerResult callBack){
        String url=Constants.Server.versionUrl;
        Get(url, VersionModel.class,callBack);
    }

    public void saveFeedback(FeedbackModel feedback,ServerResult callBack) {
        String url = getFeedbacksURL(0);
        JSONObject jObj=JSONMapper.toJSON(feedback, FeedbackModel.class);
        Post(url,jObj,FeedbackModel.class ,callBack);
    }

    private String getRestroomsURL(String route, Double latitude, Double longitude, Integer distance){
        //return "http://www.android.com/";
        String returnUrl = Constants.Server.restroomsUrl +  "?routeAlpha=" + route + "&direction=&lat=" + latitude.toString() + "&longt=" + longitude.toString();

        if(distance != null) {
            return returnUrl + "&distance=" + distance;
        } else {
            return returnUrl;
        }
    }
    private String getRestroomURL(int restroomId){
        return Constants.Server.restroomsUrl +  "/" + restroomId;
    }

    private String getRoutesURL() {
        return (Constants.Server.transitApiUrl) + "/routes/?token=" + (Constants.Server.transitApiToken);
    }

    private String getFeedbacksURL(int restStopId){
        if (restStopId != 0){
            return Constants.Server.feedbackUrl + "?restStopId=" + restStopId;
        }
        return Constants.Server.feedbackUrl + "/";
    }


}

File: RestroomFinder/Native/Apple/Restroom/ServerAPI.swift
//
//  ServerAPI.swift
//  Restroom
//
//  Created by DevTeam on 12/18/15.
//  Copyright © 2015 DevTeam. All rights reserved.
//

import Foundation
import UIKit
import MapKit
import SharedFramework

class ServerAPI: ServerBaseAPI {
    //let url:String="http://10.1.200.244/ToolBeltAPI/api/RestStops"
        
    func getRestStops(_ route: String, location: CLLocation, callBack: @escaping ([RestStopModel], NSError?) -> Void){
        let lat=location.coordinate.latitude
        let long=location.coordinate.longitude
        var r = route
        if (r == "" || r == "-1" || r == "All"){
            r = ""
        }
        let surl=getRestURL(r,lat:String(lat), long:String(long),distance:80000)
        NSLog("surl: %@", surl)
        let res = super.get(surl) as AjaxPromise<[RestStopModel]>
        res.always.addHandler({p in
            callBack(p.data!,p.error)
        })
    }
    
    func getRestStop(_ restStopId: Int, callBack: @escaping (RestStopModel, NSError?) -> Void){
        let surl=getRestURL(restStopId)
        let res = super.get(surl) as AjaxPromise<[RestStopModel]>
        res.always.addHandler({p in
            var data: RestStopModel!
            if (p.data.count>0){
                data = p.data[0]
            }
            callBack(data,p.error)
        })


    }
 
    //NOTE: New method to retrieve demo-hard-coded data for the General Office and Route 1 restrooms.
    func getRestroomsAroundGeneralOffice(_ callBack: @escaping ([RestStopModel], NSError?) -> Void) {
        let surl = getRestURL("1", lat: "37.805413", long: "-122.268528", distance: nil)
        let res = super.get(surl) as AjaxPromise<[RestStopModel]>
        res.always.addHandler({p in
            callBack(p.data,p.error)
        })
        
    }
    
    func getOperationAsync(_ badge: String, agreed: Bool, validating: Bool, callBack: @escaping (OperationModel?, NSError?) -> Void){
        let surl = Constants.Server.operationUrl
        let obj = OperationModel()
        obj.badge = badge
        obj.agreed = agreed
        obj.validating = validating
        obj.incidentDateTime = DateTime.now().toString()
        let res = super.post(surl, jsonObject: obj.toJsonObject()) as AjaxPromise<OperationModel>
        res.always.addHandler({p in
            callBack(p.data,p.error)
        })
    }
    
    func getTripPattern(_ route: String, tripId: Int, callBack: @escaping ([Timepoint], NSError?) -> Void) {
        let surl = getTripPatternURL(route, tripId: tripId)
        let res = super.get(surl) as AjaxPromise<[Timepoint]>
        res.always.addHandler({p in
            callBack(p.data,p.error)
        })
        
    }
    
    
    func getRoutes(_ callBack: @escaping ([RouteModel], NSError?) -> Void) {
        let surl = getRoutesURL()
        let res = super.get(surl) as AjaxPromise<[RouteModel]>
        res.always.addHandler({p in
            callBack(p.data,p.error)
        })
    }
    func getVersion(_ callBack: @escaping ([VersionModel], NSError?) -> Void) {
        let surl = Constants.Server.versionUrl
        let res = super.get(surl) as AjaxPromise<[VersionModel]>
        res.always.addHandler({p in
            callBack(p.data,p.error)
        })
    }
    
    func getFeedbacks(_ restStopId: Int, callBack: @escaping ([FeedbackModel], NSError?) -> Void) {
        let surl = getFeedbacksURL(restStopId)
        let res = super.get(surl) as AjaxPromise<[FeedbackModel]>
        res.always.addHandler({p in
            callBack(p.data,p.error)
        })
    }

    func saveFeedback(_ feedback: FeedbackModel, callBack: @escaping (FeedbackModel?, NSError?) -> Void) {
        let surl = getFeedbacksURL(nil)
        let res = post(surl, jsonObject: feedback.toJsonObject()) as AjaxPromise<FeedbackModel>
        res.always.addHandler({p in
            callBack(p.data,p.error)
        })
    }

    func getTripPatternURL(_ route:String, tripId:Int) -> String{
        let tripPatternURL = "\(Constants.Server.transitApiUrl)/route/\(route)/trip/\(tripId)/pattern?token=\(Constants.Server.transitApiToken)"
        print(tripPatternURL)
        return tripPatternURL
    }
    
    //NOTE: Made distance nullable
    func getRestURL(_ route:String, lat:String, long:String, distance:Int?)-> String{
        let returnUrl = "\(Constants.Server.restUrl)?routeAlpha=\(route)&direction=&lat=\(lat)&longt=\(long)"

        if(distance != nil) {
            return "\(returnUrl)&distance=\(distance)";
        } else {
            return returnUrl;
        }
    }
    func getRestURL(_ restStopId: Int)-> String{
        return  "\(Constants.Server.restUrl)/\(restStopId)"
    }
    func getOperationURL(_ badge:String)-> String{
        return "\(Constants.Server.operationUrl)?badgeNumber=\(badge)"
    }
    
    func getRoutesURL() -> String{
        return "\(Constants.Server.transitApiUrl)/routes/?token=\(Constants.Server.transitApiToken)"
    }
    func getFeedbacksURL(_ restStopId : Int?) -> String{
        if (restStopId != nil){
            return "\(Constants.Server.baseRestroomURL)/Feedback?restStopId=\(restStopId!)"
        }
        return "\(Constants.Server.baseRestroomURL)/Feedback/"
    }
    
}



